### PR TITLE
fix(metro-resolver): package exports array root shorthand

### PR DESCRIPTION
**Pull request by jbroma**
## Summary

This PR deals with array root shorthand edge case from #1236 

- [x] - fixed array root shorthand spec in tests to match Node.js resolution algorithm
- [x] - aligned implementation of package exports for array root shorthand 
- [x] - added extra `[nonstrict]` test case to highlight that non-declared paths from array should not work normally 

Changelog: [Experimental] Fix package exports array root shorthand resolution to behave like node (edge case)

## Test plan
- [x] - all tests pass

### CODE REVIEW

File: packages/metro-resolver/src/__tests__/package-exports-test.js [hunk 1]
```diff
@@ -255,6 +255,51 @@ describe('with package exports resolution enabled', () => {
     });
   });
 
+  describe('array root shorthand', () => {
```
Comment (huntie) [inline packages/metro-resolver/src/__tests__/package-exports-test.js:258]: Can we group this `describe` group under `describe('main entry point', () => {`? Will make the "should use first valid value" test more understandable in context.

**Review by huntie — COMMENTED** [review]
Thanks! I'm still trying to find where the previous root shorthand behaviour was documented. Lining up with current Node makes sense 👍🏻

File: packages/metro-resolver/src/__tests__/package-exports-test.js [hunk 2]
```diff
@@ -255,6 +255,51 @@ describe('with package exports resolution enabled', () => {
...
+    });
+
+    test('[nonstrict] should fall back and log warning when subpath is not listed', () => {
+      expect(Resolver.resolve(context, 'test-pkg/index.js', null)).toEqual({
+        type: 'sourceFile',
+        filePath: '/root/node_modules/test-pkg/index.js',
+      });
+      expect(logWarning).toHaveBeenCalledTimes(1);
+      expect(logWarning.mock.calls[0][0]).toMatchInlineSnapshot(
+        `"Attempted to import the module \\"/root/node_modules/test-pkg/index.js\\" which is not listed in the \\"exports\\" of \\"/root/node_modules/test-pkg\\". Falling back to file-based resolution. Consider updating the call site or asking the package maintainer(s) to expose this API."`,
```
[the remark this answers is not in the document]
Comment (jbroma) [reply packages/metro-resolver/src/__tests__/package-exports-test.js:289]: yes, if package exports are defined then import like the one above should not work, see the output from repro [here](https://app.warp.dev/block/LCzgyHtD0J1LHaQfLd7HmD)

File: packages/metro-resolver/src/__tests__/package-exports-test.js (hunk 1 above)
Comment (jbroma) [reply packages/metro-resolver/src/__tests__/package-exports-test.js]: I've slightly modified the structure, it made more sense to move the `[nonstrict]` test to the rest of them under `main entry point`. I've also moved the `describe` group as suggested above 👍

File: packages/metro-resolver/src/__tests__/package-exports-test.js [hunk 3]
```diff
@@ -211,6 +211,40 @@ describe('with package exports resolution enabled', () => {
       `);
     });
 
+    test('[nonstrict] should fall back to file resolution and log warning when subpath is not listed', () => {
```
Comment (huntie) [inline packages/metro-resolver/src/__tests__/package-exports-test.js:214]: nit: Ideally, can we split this new test into a separate PR? Or convince me why it belongs here :)

Comment (jbroma) [reply packages/metro-resolver/src/__tests__/package-exports-test.js]: hmm this is a test that I started implementation with. It was checking for a behaviour that was not spec compliant. I'm not sure if it makes sense to keep it at all now.

I've removed it from this PR and can add it inside of a separate PR if you want, up to you :)

**Review by huntie — APPROVED** [review]

**Comment by jbroma** [comment]
@robhogan oh I absolutely agree - there is more to this than what meets the eye :)

I think we should proceed with merging this edge case and handle other scenarios in separate issues / PR's.
The normalisation that you've provided seems like a good starting point to begin with. We can reuse [the repro](https://github.com/jbroma/metro-package-exports) from the #1236 and just add more cases there. 

Is there anything else that needs to be addressed in this PR? I see the FB internal job failed for some reason.

**Comment by jbroma** [comment]
@huntie @robhogan is there anything blocking us from proceeding with this PR?
